### PR TITLE
Fixing tutorial with DataLoader

### LINKED ISSUE #352: DataLoader tutorial does not handle num_workers > 0
I just wanted to document an issue with the tutorials https://pytorch.org/data/beta/tutorial.html#working-with-dataloader

The code in the tutorial will not work when running multiple DataLoader processes as the datapipe will be duplicated across workers:

```py
    dl = DataLoader(dataset=datapipe, batch_size=2, shuffle=True, num_workers=2)

    for i, e in enumerate(dl):
        print(e)
```
gives

```
{'label': tensor([7, 0], dtype=torch.int32), 'data': tensor([[0.5105, 0.7899],
        [0.0152, 0.5981]], dtype=torch.float64)}
{'label': tensor([7, 0], dtype=torch.int32), 'data': tensor([[0.5105, 0.7899],
        [0.0152, 0.5981]], dtype=torch.float64)}
{'label': tensor([4, 6], dtype=torch.int32), 'data': tensor([[0.9998, 0.5452],
        [0.8515, 0.8264]], dtype=torch.float64)}
{'label': tensor([4, 6], dtype=torch.int32), 'data': tensor([[0.9998, 0.5452],
        [0.8515, 0.8264]], dtype=torch.float64)}
{'label': tensor([1, 9], dtype=torch.int32), 'data': tensor([[0.8423, 0.3664],
        [0.6397, 0.6408]], dtype=torch.float64)}
{'label': tensor([1, 9], dtype=torch.int32), 'data': tensor([[0.8423, 0.3664],
        [0.6397, 0.6408]], dtype=torch.float64)}
...
```

Even though this is still beta, it may still be worth  letting users know about such pitfalls.

Also, since there are various ways to achieve the sharding, it could be useful to settle on a definite canonical way of handling all this.

### PR DESCRIPTION
**Pull request by NivekT**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #458

Updating the tutorial and README with more relevant/correct information. Minor fix to one part of `MapDataPipe` documentation as well.

Fixes #352 

Differential Revision: [D36645515](https://our.internmc.facebook.com/intern/diff/D36645515)

### CODE REVIEW

**Comment by NivekT** [comment]
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D36645515).

**Comment by NivekT** [comment]
Here is the rendered page:

Screen Shot 2022-05-24 at 6 15 27 PM
Screen Shot 2022-05-24 at 6 15 44 PM
Screen Shot 2022-05-24 at 6 15 50 PM

File: README.md [hunk 1]
```diff
@@ -177,17 +177,30 @@ Q: What should I do if the existing set of DataPipes does not do what I need?
...
 believe your use case is common enough such that the community can benefit from having your custom DataPipe added to
-this library, feel free to open a GitHub issue.
+this library, feel free to open a GitHub issue. We will be happy to discuss!
 
-Q: What happens when the `Shuffler`/`Batcher` DataPipes are used with DataLoader?
+Q: What happens when the `Shuffler` DataPipe is used with DataLoader?
 
-A: If you choose those DataPipes while setting `shuffle=True`/`batch_size>1` for DataLoader, your samples will be
-shuffled/batched more than once. You should choose one or the other.
+A. In order to enable shuffling, you need to add a `Shuffler` to your DataPipe line and set `shuffle=True` within
```
Comment (ejguan) [inline README.md:184]: Users don't necessarily set `shuffle=True` within `DataLoader` to enable `shuffle`. As long as users don't specify `shuffle=False` or not apply `shuffle_dp. set_shuffle(False)`, `DataLoader` would enable shuffle by default.

File: docs/source/tutorial.rst [hunk 2]
```diff
@@ -80,23 +80,33 @@ For this example, we will first have a helper function that generates some CSV f
...
         datapipe = dp.iter.FileLister(root_dir)
-        datapipe = datapipe.filter(filter_fn=lambda filename: "sample_data" in filename and filename.endswith(".csv"))
-        datapipe = dp.iter.FileOpener(datapipe, mode='rt')
+        datapipe = datapipe.filter(filter_fn=filter_for_data)
+        datapipe = datapipe.open_files(mode='rt')
         datapipe = datapipe.parse_csv(delimiter=",", skip_lines=1)
-        datapipe = datapipe.map(lambda row: {"label": np.array(row[0], np.int32),
-                                             "data": np.array(row[1:], dtype=np.float64)})
+        datapipe = datapipe.map(row_processer)
+        # We will also have to set `shuffle=True` later in the DataLoader
```
Comment (ejguan) [inline docs/source/tutorial.rst:103]: See comment above.

File: README.md [hunk 3]
```diff
@@ -177,17 +177,30 @@ Q: What should I do if the existing set of DataPipes does not do what I need?
...
+A: By design, there are fewer `MapDataPipes` than `IterDataPipes` to avoid duplicate implementations of the same
+functionalities as `MapDataPipe`. We encourage users to use the built-in `IterDataPipe` for various functionalities, and
+convert it to `MapDataPipe` as needed.
 
 Q: How is multiprocessing handled with DataPipes?
 
 A: Multi-process data loading is still handled by DataLoader, see the
 [DataLoader documentation for more details](https://pytorch.org/docs/stable/data.html#single-and-multi-process-data-loading).
+If you would like to shard data across processes, use `ShardingFilter` and provide a `worker_init_fn` as shown in the
+[tutorial](https://pytorch.org/data/beta/tutorial.html#working-with-dataloader).
```
Comment (ejguan) [inline README.md:203]: Could you please also add a few words about `shuffle` should be applied before `sharding_filter` to ensure all worker process has the same order of data for sharding.

**Review by ejguan — COMMENTED** [review]
Thank you so much for adding this detailed documents for users. I do have a few comments below.

**Review by ejguan — APPROVED** [review]
Thank you so much for your work. This looks amazing.

**Comment by NivekT** [comment]
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D36645515).

**Comment by Sino-Huang** [comment]
hello, the new tutorial miss the `worker_init_fn` part.
please check

**Comment by NivekT** [comment]
> hello, the new tutorial miss the `worker_init_fn` part. please check

Hi, that part has been removed because we have made changes for the next release (0.4.0, PyTorch Core 1.12.0), such that passing in `work_init_fn` for sharding should no longer be needed. Let me know if you have any question.
